### DISCUSSION TITLE
RedundantCondition and superglobals when saved to variables

### DISCUSSION THREAD

**Discussion by natecraddock** [vimeo/psalm#10474] [Q&A] [answered]
I'm working on enabling RedundantCondition checks and I've run into an issue with superglobals, specifically `$_GET` and `$_POST` that confuses me. Here is an example: https://psalm.dev/r/fd3d80222b.

In the example, psalm is reporting that the value of `$value` is never null. I'm curious why assigning one of the values of a superglobal like `$_GET` to a variable makes psalm think it cannot be null. Yet, when the superglobal key is accessed directly in an `isset()` function, there isn't any redundant condition.

As my example shows, there are some workarounds that I can use. I expect I am just not understanding how psalm handles superglobals. I would appreciate some clarification on why psalm behaves this way. Or if there are other suggestions on how to handle this case that would be appreciated.

In trying to see if this has been discussed before, it looks like the type of superglobals was changed a little over a year ago here: https://github.com/vimeo/psalm/pull/8473. Does there need to be more work done here to handle the possibly null cases?

**Comment by robchett** [comment, marked as answer]
My understanding is that $_GET and $_POST would not be populated with null value, the closest being an '' in the case of ?value or ?value= hence the types being defined as they are.

In the second example isset is acting as an array_key_exists, not actually checking for a null value.
In the third the null coalesce provides the null value in the case of the key not existing and so it's doing the same thing as the second indirectly.

It's possible to set a value yourself which is null, but Psalm will report it at that point and continue to use the hinted type elsewhere. In this case Psalm is trying to provide safety for the 99%, and modifying superglobals is frowned upon

I don't think your workarounds are workarounds, they are better code. as they won't trigger any undefined array key warnings.
If you are expecting null values in your superglobals then maybe an array_key_exists check would be better.

Hope that clears things up!

**Reply by natecraddock** [reply]
Thanks! That does clear things up. I think my confusion lies in me mixing up undefined array keys with expecting an undefined array key to be null.
